Support layers from external URLs

jstarks opened the pull request:
This change adds support for layers that are distributed from external URLs. This is specifically necessary to support downloading Windows base layers from Microsoft servers, since only Microsoft is allowed to distribute them.

Essentially this just changes pull to know how to download these layers, push to know not to upload these layers, and save and load to maintain the appropriate metadata.

Currently this is only enabled on Windows builds.

In the future we may want to extend this to support treating these foreign layers as normal layers for air-gapped scenarios. This is not supported yet.

cc @stevvooe, @aaronlehmann

jstarks committed:
Revendor docker/distribution

Signed-off-by: John Starks <jostarks@microsoft.com>

aaronlehmann commented on layer/filestore.go:
```diff
@@ -191,6 +201,23 @@ func (fms *fileMetadataStore) GetCacheID(layer ChainID) (string, error) {
+	if err != nil {
+		return nil, err
+	}
+	return ref, err
```
If the content of `foreign-source` somehow got set to `null`, this could return a nil pointer. I think it would be a little safer to have `var ref distribution.Descriptor` and unmarshal into that. Then this function could return `distribution.Descriptor` instead of `*distribution.Descriptor` so there's not even any ambiguity about whether the result can be nil.

jstarks replied:
We return nil to mean that there is no foreign source. I think this is the right way to handle it, rather than requiring callers to check for a non-empty digest or return an additional bool or something.

You're right, though; if `foreign-source` exists, we should return a non-nil value.

aaronlehmann replied:
I think it could be confusing for the function to return a `nil` error but also a `nil` result. The general convention is that `err == nil` means there's a valid value being returned.

jstarks replied:
How do you suggest indicating to the caller that there is no foreign source?

aaronlehmann replied:
I would probably return a sentinel error value like `ErrNoForeignSource` to return in this case.

aaronlehmann commented on layer/filestore.go:
```diff
@@ -98,6 +100,14 @@ func (fm *fileMetadataTransaction) SetCacheID(cacheID string) error {
 	return ioutil.WriteFile(filepath.Join(fm.root, "cache-id"), []byte(cacheID), 0644)
 }
 
+func (fm *fileMetadataTransaction) SetForeignSource(ref *distribution.Descriptor) error {
```
Basically cosmetic, but I think in most existing code `distribution.Descriptor` is passed by value instead of by reference.

jstarks replied:
Good idea. This means that you can't clear the foreign source with this, but this code didn't handle this case anyway.

aaronlehmann commented on distribution/push_v2.go:
```diff
@@ -240,6 +240,11 @@ func (pd *v2PushDescriptor) DiffID() layer.DiffID {
 
 func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.Output) (distribution.Descriptor, error) {
+	if d := pd.layer.ForeignSource(); d != nil {
+		progress.Update(progressOutput, pd.ID(), "Layer already exists")
```
I'm wondering if it's better to use different wording indicating that the layer is being skipped before it's a foreign source layer. It could certainly help in debugging to know that this is a not an ordinary layer and it's not expected to exist on the registry after a push.

jstarks replied:
Will fix.

aaronlehmann commented on distribution/pull_v2.go:
```diff
@@ -501,6 +519,29 @@ func (p *v2Puller) pullSchema2(ctx context.Context, ref reference.Named, mfst *s
+		}
+
+		if d.MediaType == schema2.MediaTypeForeignLayer && len(d.URLs) > 0 {
+			if runtime.GOOS != "windows" {
```
Rather than the push/pull code looking at OS directly, I'd suggest having a function in the `layer` package like `ForeignSourceSupported`.

Another way to do it that could be interesting is to remove `ForeignSource` from the `Layer` interface, and only define the method on Windows (by putting it in a `*_windows.go` file). Then we could put it in its own interface definition:

```go
type ForeignSourcer interface {
	// ForeignSource returns the foreign reference information for
	// the layer.
	ForeignSource() *distribution.Descriptor
}
```

Then the download manager could do something like:

```go
_, foreignSrcSupported := ldm.layerStore.(layer.ForeignSourcer)
if descriptor.foreignSrc != nil && !foreignSourceSupported {
        return errors.New("foreign layers are not supported on this OS")
}
```

The advantage of doing it this way is that we wouldn't pollute the `Layer` interface, so other OS' never have to implement it or deal with it, and things like mocks in unit tests don't need to implement the `ForeignSource` method.

stevvooe replied:
I'm a little confused as to why we care whether or not it is foreign at this point. Shouldn't we just honor the URL if it is available?

jstarks replied:
Hmm. `ForeignSourcer` would be implemented by `roLayer`, but we don't have an `roLayer` at the point we're trying to determine if this is supported. So I'll probably have to go with your first suggestion.

@stevvooe I had received feedback that we should limit this capability to Windows, i.e. we don't pay attention to the URL on Linux.

stevvooe replied:
@jstarks I think we can limit the ability to windows (by blanking out urls on non-foreign, descriptors) without introducing the concept of foreign source this deeply into the code. I'm worried this abstraction is leaking too widely.

runcom replied:
sorry, but why this has to be windows specific? can others take advantage of this by putting base layers when appropriate on different registries?

runcom replied:
I think this is awesome for anyone willing to host their base layers somewhere else than the docker hub or any private registry.

stevvooe replied:
> I think this is awesome for anyone willing to host their base layers somewhere else than the docker hub or any private registry.

I agree but that is no reason to hold up this PR. My point is that we need to be careful to not leak the foreign source abstraction, since it will make it more challenging to bring wider support here in the future. In other words, having foreign source here will make it harder to support this widely.

runcom replied:
I see, I don't see why designing this having in mind that it's going to only work on Windows. I agree we need to be careful in designing this though

stevvooe replied:
The focus on this PR should be ensuring that decisions made here don't prevent or complicate future additions along these lines. Please help here, as it will make widening the scope of this feature easier in the future.

stevvooe commented on distribution/pull_v2.go:
```diff
@@ -182,13 +188,25 @@ func (ld *v2LayerDescriptor) Download(ctx context.Context, progressOutput progre
-		if err == distribution.ErrBlobUnknown {
-			return nil, 0, xfer.DoNotRetry{Err: err}
+	var layerDownload distribution.ReadSeekCloser
+	if ld.foreignSrc == nil {
```
Perhaps this becomes a method `(*v2LayerDescriptor).open`.

aaronlehmann commented on distribution/pull_v2.go:
```diff
@@ -182,13 +188,25 @@ func (ld *v2LayerDescriptor) Download(ctx context.Context, progressOutput progre
+			_, err = layerDownload.Seek(0, os.SEEK_SET)
+			if err == nil {
+				break
+			}
```
Call `layerDownload.Close()` here to avoid leaking connections associated with unsuccessful attempts.

jstarks replied:
Good catch

stevvooe commented on distribution/pull_v2.go:
```diff
@@ -147,6 +149,10 @@ func (ld *v2LayerDescriptor) DiffID() (layer.DiffID, error) {
 }
 
+func (ld *v2LayerDescriptor) ForeignSource() *distribution.Descriptor {
+	return ld.foreignSrc
```
At this point, why does it matter if it is a foreign source? Shouldn't we just have a URL?

jstarks replied:
We need to remember its descriptor so that we can construct a manifest for this image later. `ForeignSource()` is used to retrieve that descriptor when registering the layer.

stevvooe replied:
Would all layers have a `Descriptor`?

jstarks replied:
As currently designed, only layers that got a descriptor from `docker load` or had external URLs from `docker pull` will store a descriptor.

stevvooe replied:
Great!

See https://github.com/docker/docker/pull/22866#discussion_r64658849. I only think we need to ensure that we don't have to do too much work to get this `Descriptors` working for other layers.

stevvooe commented on distribution/xfer/download.go:
```diff
@@ -62,6 +63,9 @@ type DownloadDescriptor interface {
 	DiffID() (layer.DiffID, error)
+	// ForeignSource returns nil or the foreign layer's information and
+	// source.
+	ForeignSource() *distribution.Descriptor
```
Why not just expose a descriptor for the layer and then have method to indicate whether or not it is foreign?

jstarks replied:
Do we want to bother storing a descriptor in non-foreign cases? That might be tricky since we don't know what the digest is going to be until we actually construct and compress the layer tar. It seemed simpler to only store the descriptor in the foreign case.

stevvooe replied:
That might be a good invariant to enforce. The lack of digests on existing layers is an ongoing problem. I'd hate to exacerbate this problem by sending the message that only window's foreign layers have a digest.

aaronlehmann replied:
We already store digests for all layers. See https://github.com/docker/docker/blob/master/distribution/metadata/v2_metadata_service.go

stevvooe replied:
Then let's store this there. There is no reason to do this twice.

aaronlehmann replied:
I've thought about this a bit previously, but the issue is that distribution metadata only tracks layers that are pulled or push. We need to preserve the foreign source information for layers brought in with `docker load` as well. It's technically possible to use the distribution metadata, but it would leak this metadata out of the context of push/pull.

stevvooe replied:
@jstarks After discussing this with @aaronlehmann, we thought it would be okay as long as the ondisk formats would allow us to port this to other platforms. I think the only change that needs to be made is to call name the json file `descriptor.json` instead of `foreign-source.json`.

AkihiroSuda commented on distribution/pull_v2.go:
```diff
@@ -501,6 +519,29 @@ func (p *v2Puller) pullSchema2(ctx context.Context, ref reference.Named, mfst *s
+
+		if d.MediaType == schema2.MediaTypeForeignLayer && len(d.URLs) > 0 {
+			if runtime.GOOS != "windows" {
+				return "", "", errors.New("foreign layers are not supported on this OS")
```
I suggest making the error more descriptive (non-binding): e.g. `errors.New("foreign layers are not supported on this OS (currently only supported on Windows)")`

runcom commented:
> Currently this is only enabled on Windows builds.
> 
> In the future we may want to extend this to support treating these foreign layers as normal layers for air-gapped scenarios. This is not supported yet.

@jstarks is there anything preventing _right_ _now_ to enable this on linux for instance?

rhatdan commented:
We definitely want something like this for supporting RHEL base images.

jstarks commented:
@runcom No, this could work on any OS. I had received some feedback that we might want to limit it to Windows initially.

jstarks committed:
Support layers from external URLs

This is used to support downloading Windows base images from Microsoft
servers.

Signed-off-by: John Starks <jostarks@microsoft.com>

jstarks commented:
I've pushed an update that fixes a few issues, renames `foreign-source` to `descriptor.json` and separates more of this out to only build on Windows so that we don't entangle this too much into the rest of Docker. There is more separating that could be done (e.g. move some of the layer store code to _windows files), but I'm not sure it's really worth it.

Note, I haven't validated these changes yet.

aaronlehmann commented:
LGTM pending your validation

runcom commented:
> @runcom No, this could work on any OS. I had received some feedback that we might want to limit it to Windows initially.

:/

miminar commented:
+1 for this. +2 for making this OS generic. It would in fact simplify the PR.

llunved commented:
This is something we really need on Linux as much as Windows.

jstarks commented:
@aaronlehmann I've validated these changes locally.

jstarks commented:
I don't think the CI failures are related to this change.

dmcgowan commented:
This comment applies to this PR as well https://github.com/docker/docker/pull/23014#issuecomment-222005512. The change is only intended as a code interface and not a change to what is stored on disk. I could support getting this in as in with the understanding the code will be refactored.

stevvooe commented:
LGTM

taylorb-microsoft commented:
Can we merge this?

Tagging John/Stefan @jhowardmsft @swernli while john is out of the office.

icecrime commented:
I trust @aaronlehmann and @stevvooe on the review, I only gave it a read.

LGTM :+1:

icecrime commented:
I don't think there's any documentation involved on docker side, but we might need to update specs in the distribution repository maybe?

runcom commented on image/tarexport/tarexport.go:
```diff
@@ -15,10 +16,11 @@ const (
+	RepoTags     []string
+	Layers       []string
+	Parent       image.ID                                  `json:",omitempty"`
+	LayerSources map[layer.DiffID]*distribution.Descriptor `json:",omitempty"`
```
it's unfortunate this map is indexed by `layer.DiffID` - this basically prevents having common code when integrating OCI save/load (#26369).
You can't get foreign sources _by diffID_ in OCI since all hashes are just from compressed layers digests and at the time we build this `manifestItem` for an OCI tarball we don't have info about diffIDs and cannot populate this map.

@aaronlehmann @stevvooe @tonistiigi any idea on how to convert this to something more useful to be used in the OCI PR also?

aaronlehmann replied:
I suppose you could use `digest.Digest`, since both `DiffID`s and OCI foreign sources are digests of some sort?

runcom replied:
mmm nope, it's a matter of the different digest actually. the index in LayerSources are non-compressed digests, OCI has compressed digests
